runtime/near-vm-logic/src/logic.rs

```diff
@@ -2454,28 +2454,35 @@ impl<'a> VMLogic<'a> {
             .into());
         }
         self.gas_counter.pay_per(storage_read_key_byte, key.len() as u64)?;
-        let nodes_before = self.ext.get_trie_nodes_count();
-        let read_mode = if cfg!(feature = "protocol_feature_flat_state") {
-            StorageGetMode::FlatStorage
+        let (value_ptr, tn_db_reads, tn_mem_reads) = if checked_feature!(
+            "protocol_feature_flat_state",
+            FlatStorageReads,
+            self.current_protocol_version
+        ) {
+            (self.ext.storage_get(&key, StorageGetMode::FlatStorage), 0, 0)
         } else {
-            StorageGetMode::Trie
+            let nodes_before = self.ext.get_trie_nodes_count();
+            let value_ptr = self.ext.storage_get(&key, StorageGetMode::Trie);
+            let nodes_delta = self
+                .ext
+                .get_trie_nodes_count()
+                .checked_sub(&nodes_before)
+                .ok_or(InconsistentStateError::IntegerOverflow)?;
+            self.gas_counter.add_trie_fees(&nodes_delta)?;
+            (value_ptr, nodes_delta.db_reads, nodes_delta.mem_reads)
         };
-        let read = self.ext.storage_get(&key, read_mode);
-        let nodes_delta = self
-            .ext
-            .get_trie_nodes_count()
-            .checked_sub(&nodes_before)
-            .ok_or(InconsistentStateError::IntegerOverflow)?;
-        self.gas_counter.add_trie_fees(&nodes_delta)?;
-        let read = Self::deref_value(&mut self.gas_counter, storage_read_value_byte, read?)?;
+
+        let read = Self::deref_value(&mut self.gas_counter, storage_read_value_byte, value_ptr?)?;
 
         near_o11y::io_trace!(
             storage_op = "read",
             key = %near_o11y::pretty::Bytes(&key),
             size = read.as_ref().map(Vec::len),
-            tn_db_reads = nodes_delta.db_reads,
-            tn_mem_reads = nodes_delta.mem_reads,
+            tn_db_reads = tn_db_reads,
+            tn_mem_reads = tn_mem_reads,
```

Comment:
I think this output for the IO trace is now wrong if `self.ext.storage_get(&key, StorageGetMode::FlatStorage)` falls back to the trie. This messes up the statistics that the IO traces are most useful for (hit rates etc). Can we please keep this printing the real DB access values?